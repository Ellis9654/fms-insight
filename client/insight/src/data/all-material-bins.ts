/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import * as im from "immutable";

export type MaterialList = ReadonlyArray<Readonly<api.IInProcessMaterial>>;

export type AllMaterialBins = im.Map<string, MaterialList>;

export function selectAllMaterialIntoBins(curSt: Readonly<api.ICurrentStatus>): AllMaterialBins {
  const palLoc = im
    .Map(curSt.pallets)
    .map(st => " (" + st.currentPalletLocation.group + " #" + st.currentPalletLocation.num.toString() + ")");

  return im
    .Seq(curSt.material)
    .map(mat => {
      switch (mat.location.type) {
        case api.LocType.InQueue:
          return { region: mat.location.currentQueue || "Queue", mat };
        case api.LocType.OnPallet:
          let region = "Pallet";
          if (mat.location.pallet) {
            region = "Pallet " + mat.location.pallet.toString() + palLoc.get(mat.location.pallet) || "";
          }
          return { region, mat };
        case api.LocType.Free:
        default:
          switch (mat.action.type) {
            case api.ActionType.Loading:
              return { region: "Raw Material", mat };
            default:
              return { region: "Free Material", mat };
          }
      }
    })
    .groupBy(x => x.region)
    .map(group =>
      group
        .map(x => x.mat)
        .sortBy(mat => {
          switch (mat.location.type) {
            case api.LocType.OnPallet:
              return mat.location.face;
            case api.LocType.InQueue:
            case api.LocType.Free:
            default:
              return mat.location.queuePosition;
          }
        })
        .valueSeq()
        .toArray()
    )
    .toMap();
}
