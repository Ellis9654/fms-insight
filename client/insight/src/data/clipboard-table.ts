/* Copyright (c) 2019, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { LazySeq } from "./lazyseq";
import { fieldsHashCode } from "prelude-ts";
import { FilteredStationCycles, stat_name_and_num } from "./events.cycles";
import { format } from "date-fns";
const copy = require("copy-to-clipboard");

export interface ClipboardTablePoint {
  readonly x: Date;
  readonly y: string;
  readonly label: string;
}

class PointsTableCell {
  public constructor(public readonly x: number, public readonly y: string) {}
  equals(other: PointsTableCell): boolean {
    return this.x === other.x && this.y === other.y;
  }
  hashCode(): number {
    return fieldsHashCode(this.x, this.y);
  }
  toString(): string {
    return `{x: ${new Date(this.x).toISOString()}, y: ${this.y}}`;
  }
}

export function buildPointsTable(yTitle: string, points: ReadonlyArray<ClipboardTablePoint>): string {
  const cells = LazySeq.ofIterable(points).toMap(
    p => [new PointsTableCell(p.x.getTime(), p.y), p],
    (_, c) => c // cells should be unique, but just in case take the second
  );
  const days = LazySeq.ofIterable(points)
    .toSet(p => p.x.getTime())
    .toArray({ sortOn: x => x });
  const rows = LazySeq.ofIterable(points)
    .toSet(p => p.y)
    .toArray({ sortOn: x => x });

  let table = "<table>\n<thead><tr><th>" + yTitle + "</th>";
  for (let x of days) {
    table += "<th>" + new Date(x).toDateString() + "</th>";
  }
  table += "</tr></thead>\n<tbody>\n";
  for (let y of rows) {
    table += "<tr><th>" + y + "</th>";
    for (let x of days) {
      const cell = cells.get(new PointsTableCell(x, y));
      if (cell.isSome()) {
        table += "<td>" + cell.get().label + "</td>";
      } else {
        table += "<td></td>";
      }
    }
    table += "</tr>\n";
  }
  table += "</tbody>\n</table>";
  return table;
}

export function copyPointsToClipboard(yTitle: string, points: ReadonlyArray<ClipboardTablePoint>): void {
  copy(buildPointsTable(yTitle, points));
}

export function buildCycleTable(cycles: FilteredStationCycles): string {
  let table = "<table>\n<thead><tr>";
  table += "<th>Date</th><th>Part</th><th>Station</th><th>Pallet</th>";
  table += "<th>Serial</th><th>Workorder</th><td>Elapsed Min</td><td>Active Min</td>";
  table += "</tr></thead>\n<tbody>\n";
  for (let cycle of LazySeq.ofIterable(cycles.data).flatMap(([_, c]) => c)) {
    table += "<tr>";
    table += "<td>" + format(cycle.x, "MMM D, YYYY, H:mm a") + "</td>";
    table += "<td>" + cycle.part + "-" + cycle.process.toString() + "</td>";
    table += "<td>" + stat_name_and_num(cycle.stationGroup, cycle.stationNumber) + "</td>";
    table += "<td>" + cycle.pallet + "</td>";
    table += "<td>" + (cycle.serial || "") + "</td>";
    table += "<td>" + (cycle.workorder || "") + "</td>";
    table += "<td>" + cycle.y.toFixed(1) + "</td>";
    table += "<td>" + cycle.active.toFixed(1) + "</td>";
    table += "</tr>\n";
  }
  table += "</tbody>\n</table>";
  return table;
}

export function copyCyclesToClipboard(cycles: FilteredStationCycles): void {
  copy(buildCycleTable(cycles));
}
