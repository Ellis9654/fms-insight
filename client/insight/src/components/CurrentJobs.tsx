/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import * as im from 'immutable';
import { duration } from 'moment';

import * as api from '../data/api';
import { Store } from '../data/store';
import { FlexibleWidthXYPlot, HorizontalBarSeries, XAxis, YAxis, MarkSeries, VerticalGridLines } from 'react-vis';

import 'react-vis/dist/style.css';

interface DataPoint {
  part: string;
  completed: number;
  totalPlan: number;
}

export function DisplayJob(job: api.IInProcessJob, proc: number): DataPoint {
  const totalPlan = job.cyclesOnFirstProcess.reduce((a, b) => a + b, 0);
  const completed = job.completed[proc].reduce((a, b) => a + b, 0);

  const stops = job.procsAndPaths[proc].paths[0].stops;
  let cycleTime = duration();
  for (let i = 0; i < stops.length; i++) {
    const x = duration(stops[i].expectedCycleTime);
    cycleTime = cycleTime.add(x);
  }
  const cycleTimeHours = cycleTime.asHours();
  return {
    part: job.partName + '-' + (proc + 1).toString(),
    completed: cycleTimeHours * completed,
    totalPlan: cycleTimeHours * totalPlan
  };
}

export interface Props {
  jobs: ReadonlyArray<Readonly<api.IInProcessJob>>;
}

export function CurrentJobs(p: Props) {
  const data = im.Seq(p.jobs)
    .flatMap(j =>
      im.Range(0, j.procsAndPaths.length).map(proc =>
        DisplayJob(j, proc)
      )
    )
    .sortBy(pt => pt.part)
    .reverse()
    .cacheResult();
  const completedData =
    data.map((pt, i) => ({x: pt.completed, y: i, part: pt.part}))
        .toArray();
  const planData =
    data.map((pt, i) => ({x: pt.totalPlan, y: i, part: pt.part}))
        .toArray();
  return (
    <div>
      <FlexibleWidthXYPlot
          height={300}
          margin={{left: 70, right: 10, top: 10, bottom: 40}}
          yType="ordinal"
      >
        <XAxis/>
        <YAxis tickFormat={(y: number, i: number) => completedData[i].part}/>
        <VerticalGridLines color="green"/>
        <HorizontalBarSeries data={completedData}/>
        <MarkSeries data={planData}/>
      </FlexibleWidthXYPlot>
      <div style={{'textAlign': 'center'}}>Machine Hours</div>
    </div>
  );
}

export default connect(
  (s: Store) => {
    return {
        jobs: Object.values(s.Jobs.current_status.jobs)
    };
  }
)(CurrentJobs);