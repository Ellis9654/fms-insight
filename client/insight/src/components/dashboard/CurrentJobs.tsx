/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import * as im from 'immutable';
import { duration } from 'moment';
import { createSelector } from 'reselect';
import * as numerable from 'numeral';

import * as api from '../../data/api';
import { Store } from '../../data/store';
import {
  FlexibleWidthXYPlot,
  FlexibleXYPlot,
  HorizontalBarSeries,
  XAxis,
  YAxis,
  CustomSVGSeries,
  VerticalGridLines,
  HorizontalGridLines,
  Hint
} from 'react-vis';

interface DataPoint {
  readonly part: string;
  readonly completed: number;
  readonly completedCount: number;
  readonly totalPlan: number;
  readonly totalCount: number;
}

function displayJob(job: api.IInProcessJob, proc: number): DataPoint {
  const totalPlan = job.cyclesOnFirstProcess.reduce((a, b) => a + b, 0);
  const completed = job.completed[proc].reduce((a, b) => a + b, 0);

  const stops = job.procsAndPaths[proc].paths[0].stops;
  let cycleTime = duration();
  for (let i = 0; i < stops.length; i++) {
    const x = duration(stops[i].expectedCycleTime);
    cycleTime = cycleTime.add(x);
  }
  const cycleTimeHours = cycleTime.asHours();
  return {
    part: job.partName + '-' + (proc + 1).toString(),
    completed: cycleTimeHours * completed,
    completedCount: completed,
    totalPlan: cycleTimeHours * totalPlan,
    totalCount: totalPlan,
  };
}

interface CompletedDataPoint extends DataPoint {
  readonly x: number;
  readonly y: number;
}

export interface DataPoints {
  readonly completedData: ReadonlyArray<CompletedDataPoint>;
  readonly planData: ReadonlyArray<{x: number, y: number}>;
}

export function jobsToPoints(jobs: ReadonlyArray<Readonly<api.IInProcessJob>>): DataPoints {
  const points = im.Seq(jobs)
    .flatMap(j =>
      im.Range(0, j.procsAndPaths.length).map(proc =>
        displayJob(j, proc)
      )
    )
    .sortBy(pt => pt.part)
    .reverse()
    .cacheResult();
  const completedData =
    points.map((pt, i) => ({...pt, x: pt.completed, y: i}))
        .toArray();
  const planData =
    points.map((pt, i) => ({x: pt.totalPlan, y: i}))
        .toArray();
  return {completedData, planData};
}

const targetMark = () => (
  <rect x="-2" y="-5" width="4" height="10" fill="black"/>
);

interface PlotProps {
  readonly cnt: number;
  readonly children: (JSX.Element | undefined)[];
}

function FillViewportPlot({children}: PlotProps) {
  return (
    <div style={{'flexGrow': 1, 'display': 'flex', 'flexDirection': 'column'}}>
      <div style={{'flexGrow': 1, 'position': 'relative'}}>
        <div style={{'position': 'absolute', 'top': 0, 'left': 0, 'bottom': 0, 'right': 0}}>
          <FlexibleXYPlot
              margin={{left: 70, right: 10, top: 10, bottom: 40}}
              yType="ordinal"
          >
            {children}
          </FlexibleXYPlot>
        </div>
      </div>
      <div style={{'textAlign': 'center', 'marginBottom': '8px', 'color': '#6b6b76'}}>Machine Hours</div>
    </div>
  );
}

function ScrollablePlot({children, cnt}: PlotProps) {
  return (
    <div>
      <FlexibleWidthXYPlot
          height={cnt * 40}
          margin={{left: 70, right: 10, top: 10, bottom: 40}}
          yType="ordinal"
      >
        {children}
      </FlexibleWidthXYPlot>
      <div style={{'textAlign': 'center'}}>Machine Hours</div>
    </div>
  );
}

export interface Props extends DataPoints {
  readonly fillViewport: boolean;
}

interface JobState {
  readonly hoveredJob?: CompletedDataPoint;
}

function format_hint(j: CompletedDataPoint) {
  return [
    {title: "Part", value: j.part},
    {title: "Completed", value: j.completedCount},
    {title: "Planned", value: j.totalCount},
    {title: "Remaining Time", value: numerable(j.totalPlan - j.completed).format('0.0') + " hours"}
  ];
}

export class CurrentJobs extends React.PureComponent<Props, JobState> {
  state: JobState = {};

  setHint = (j: CompletedDataPoint) => {
    this.setState({hoveredJob: j});
  }

  render() {
    const Plot = this.props.fillViewport ? FillViewportPlot : ScrollablePlot;
    return (
      <Plot cnt={this.props.completedData.length}>
        <XAxis/>
        <YAxis tickFormat={(y: number, i: number) => this.props.completedData[i].part}/>
        <HorizontalGridLines/>
        <VerticalGridLines/>
        <HorizontalBarSeries
          data={this.props.completedData}
          color="#795548"
          onValueMouseOver={this.setHint}
          onValueMouseOut={() => this.setState({hoveredJob: undefined})}
        />
        <CustomSVGSeries data={this.props.planData} customComponent={targetMark}/>
        {
          this.state.hoveredJob === undefined ? undefined :
            <Hint value={this.state.hoveredJob} format={format_hint}/>
        }
      </Plot>
    );
  }
}

const jobsToPointsSelector = createSelector(
  (s: Store) => s.Current.current_status.jobs,
  js => jobsToPoints(Object.values(js))
);

export default connect(
  (s: Store) => {
    return jobsToPointsSelector(s);
  }
)(CurrentJobs);