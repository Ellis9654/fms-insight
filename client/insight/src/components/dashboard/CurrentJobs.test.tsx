/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from 'react';
import { render, cleanup } from 'react-testing-library';
afterEach(cleanup);

import { jobsToPoints, CurrentJobs } from './CurrentJobs';
import * as api from '../../data/api';

it('renders the current jobs', () => {
  const completedData = [
    {x: 10, y: 1, part: 'abc', completed: 5, completedCount: 1, totalPlan: 10, totalCount: 11},
    {x: 11, y: 2, part: 'def', completed: 1, completedCount: 6, totalPlan: 1, totalCount: 9},
    {x: 16, y: 3, part: 'zzz', completed: 0, completedCount: 0, totalPlan: 15, totalCount: 6}
  ];
  const planData = [
    {x: 16, y: 1},
    {x: 17, y: 2},
    {x: 22, y: 3}
  ];
  const {container} = render(
    <CurrentJobs completedData={completedData} planData={planData} fillViewport={false}/>);
  expect(container).toMatchSnapshot('current job graph scrollable');
});

it('renders the current jobs with filling viewport', () => {
  const completedData = [
    {x: 12, y: 1, part: 'ddd', completed: 4, completedCount: 6, totalPlan: 8, totalCount: 9},
    {x: 7, y: 2, part: 'eee', completed: 2, completedCount: 3, totalPlan: 7, totalCount: 100},
    {x: 9, y: 3, part: 'fff', completed: 1, completedCount: 8, totalPlan: 3, totalCount: 99}
  ];
  const planData = [
    {x: 40, y: 1},
    {x: 33, y: 2},
    {x: 24, y: 3}
  ];
  const {container} = render(
    <CurrentJobs completedData={completedData} planData={planData} fillViewport={true}/>);
  expect(container).toMatchSnapshot('current job graph filling viewport');
});

it('converts events to points', () => {
  const dummyHold: api.IJobHoldPattern = {
    userHold: false,
    reasonForUserHold: '',
    holdUnholdPattern: [],
    holdUnholdPatternStartUTC: new Date(),
    holdUnholdPatternRepeats: false,
  };
  const jobs: api.IInProcessJob[] = [
    {
      routeStartUTC: new Date(),
      routeEndUTC: new Date(),
      archived: false,
      copiedToSystem: true,
      partName: "part1",
      unique: "uniq1",
      priority: 10,
      manuallyCreated: false,
      createMarkingData: true,
      holdEntireJob: new api.JobHoldPattern(dummyHold),
      cyclesOnFirstProcess: [40, 42],
      completed: [[20, 21], [22]],
      procsAndPaths: [
        new api.ProcessInfo({
          paths: [
            new api.ProcPathInfo({
              pathGroup: 1,
              pallets: ["pal1"],
              load: [1],
              unload: [2],
              stops: [
                new api.JobMachiningStop({
                  stations: { "1": "progabc" },
                  tools: {},
                  stationGroup: "MC",
                  expectedCycleTime: "01:15:00",
                })
              ],
              expectedLoadTime: '00:05:00',
              expectedUnloadTime: '00:04:00',
              simulatedStartingUTC: new Date(),
              simulatedAverageFlowTime: "",
              holdMachining: new api.JobHoldPattern(dummyHold),
              holdLoadUnload: new api.JobHoldPattern(dummyHold),
              partsPerPallet: 1,
            }),
            new api.ProcPathInfo({
              pathGroup: 1,
              pallets: ["pal2"],
              load: [1],
              unload: [2],
              stops: [
                new api.JobMachiningStop({
                  stations: { "1": "progabc" },
                  tools: {},
                  stationGroup: "MC",
                  expectedCycleTime: "01:15:00",
                })
              ],
              expectedLoadTime: '00:05:00',
              expectedUnloadTime: '00:04:00',
              simulatedStartingUTC: new Date(),
              simulatedAverageFlowTime: "",
              holdMachining: new api.JobHoldPattern(dummyHold),
              holdLoadUnload: new api.JobHoldPattern(dummyHold),
              partsPerPallet: 1,
            })
          ]
        }),
        new api.ProcessInfo({
          paths: [
            new api.ProcPathInfo({
              pathGroup: 1,
              pallets: ["pal1"],
              stops: [
                new api.JobMachiningStop({
                  stations: { "1": "progabc" },
                  tools: {},
                  stationGroup: "MC",
                  expectedCycleTime: "00:45:00",
                })
              ],
              expectedLoadTime: '00:05:00',
              expectedUnloadTime: '00:04:00',
              load: [1],
              unload: [2],
              simulatedStartingUTC: new Date(),
              simulatedAverageFlowTime: "",
              holdMachining: new api.JobHoldPattern(dummyHold),
              holdLoadUnload: new api.JobHoldPattern(dummyHold),
              partsPerPallet: 1,
            })
          ]
        })
      ]
    }
  ];
  const points = jobsToPoints(jobs);
  expect(points).toMatchSnapshot('points for sample jobs');
});