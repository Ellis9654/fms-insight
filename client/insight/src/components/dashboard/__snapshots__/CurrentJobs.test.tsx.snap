// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts events to points: points for sample jobs 1`] = `
Object {
  "completedData": Array [
    Object {
      "completed": 16.5,
      "completedCount": 22,
      "part": "part1-2",
      "totalCount": 82,
      "totalPlan": 61.5,
      "x": 16.5,
      "y": 0,
    },
    Object {
      "completed": 51.25,
      "completedCount": 41,
      "part": "part1-1",
      "totalCount": 82,
      "totalPlan": 102.5,
      "x": 51.25,
      "y": 1,
    },
  ],
  "planData": Array [
    Object {
      "x": 61.5,
      "y": 0,
    },
    Object {
      "x": 102.5,
      "y": 1,
    },
  ],
}
`;

exports[`renders the current jobs with filling viewport: current job graph filling viewport 1`] = `
<FillViewportPlot
  cnt={3}
>
  <XAxis
    attr="x"
    attrAxis="y"
    orientation="bottom"
  />
  <YAxis
    attr="y"
    attrAxis="x"
    orientation="left"
    tickFormat={[Function]}
  />
  <HorizontalGridLines
    attr="y"
    direction="horizontal"
  />
  <VerticalGridLines
    attr="x"
    direction="vertical"
  />
  <HorizontalBarSeries
    className=""
    color="#795548"
    data={
      Array [
        Object {
          "completed": 4,
          "completedCount": 6,
          "part": "ddd",
          "totalCount": 9,
          "totalPlan": 8,
          "x": 12,
          "y": 1,
        },
        Object {
          "completed": 2,
          "completedCount": 3,
          "part": "eee",
          "totalCount": 100,
          "totalPlan": 7,
          "x": 7,
          "y": 2,
        },
        Object {
          "completed": 1,
          "completedCount": 8,
          "part": "fff",
          "totalCount": 99,
          "totalPlan": 3,
          "x": 9,
          "y": 3,
        },
      ]
    }
    onValueMouseOut={[Function]}
    onValueMouseOver={[Function]}
    stack={false}
    style={Object {}}
  />
  <AbstractSeries
    animation={false}
    className=""
    customComponent={[Function]}
    data={
      Array [
        Object {
          "x": 40,
          "y": 1,
        },
        Object {
          "x": 33,
          "y": 2,
        },
        Object {
          "x": 24,
          "y": 3,
        },
      ]
    }
    stack={false}
    style={Object {}}
  />
</FillViewportPlot>
`;

exports[`renders the current jobs: current job graph scrollable 1`] = `
<ScrollablePlot
  cnt={3}
>
  <XAxis
    attr="x"
    attrAxis="y"
    orientation="bottom"
  />
  <YAxis
    attr="y"
    attrAxis="x"
    orientation="left"
    tickFormat={[Function]}
  />
  <HorizontalGridLines
    attr="y"
    direction="horizontal"
  />
  <VerticalGridLines
    attr="x"
    direction="vertical"
  />
  <HorizontalBarSeries
    className=""
    color="#795548"
    data={
      Array [
        Object {
          "completed": 5,
          "completedCount": 1,
          "part": "abc",
          "totalCount": 11,
          "totalPlan": 10,
          "x": 10,
          "y": 1,
        },
        Object {
          "completed": 1,
          "completedCount": 6,
          "part": "def",
          "totalCount": 9,
          "totalPlan": 1,
          "x": 11,
          "y": 2,
        },
        Object {
          "completed": 0,
          "completedCount": 0,
          "part": "zzz",
          "totalCount": 6,
          "totalPlan": 15,
          "x": 16,
          "y": 3,
        },
      ]
    }
    onValueMouseOut={[Function]}
    onValueMouseOver={[Function]}
    stack={false}
    style={Object {}}
  />
  <AbstractSeries
    animation={false}
    className=""
    customComponent={[Function]}
    data={
      Array [
        Object {
          "x": 16,
          "y": 1,
        },
        Object {
          "x": 17,
          "y": 2,
        },
        Object {
          "x": 22,
          "y": 3,
        },
      ]
    }
    stack={false}
    style={Object {}}
  />
</ScrollablePlot>
`;

exports[`shows hint for mouse over: after mousing out of value 1`] = `
<ScrollablePlot
  cnt={3}
>
  <XAxis
    attr="x"
    attrAxis="y"
    orientation="bottom"
  />
  <YAxis
    attr="y"
    attrAxis="x"
    orientation="left"
    tickFormat={[Function]}
  />
  <HorizontalGridLines
    attr="y"
    direction="horizontal"
  />
  <VerticalGridLines
    attr="x"
    direction="vertical"
  />
  <HorizontalBarSeries
    className=""
    color="#795548"
    data={
      Array [
        Object {
          "completed": 5,
          "completedCount": 1,
          "part": "abc",
          "totalCount": 11,
          "totalPlan": 10,
          "x": 10,
          "y": 1,
        },
        Object {
          "completed": 1,
          "completedCount": 6,
          "part": "def",
          "totalCount": 9,
          "totalPlan": 1,
          "x": 11,
          "y": 2,
        },
        Object {
          "completed": 0,
          "completedCount": 0,
          "part": "zzz",
          "totalCount": 6,
          "totalPlan": 15,
          "x": 16,
          "y": 3,
        },
      ]
    }
    onValueMouseOut={[Function]}
    onValueMouseOver={[Function]}
    stack={false}
    style={Object {}}
  />
  <AbstractSeries
    animation={false}
    className=""
    customComponent={[Function]}
    data={
      Array [
        Object {
          "x": 16,
          "y": 1,
        },
        Object {
          "x": 17,
          "y": 2,
        },
        Object {
          "x": 22,
          "y": 3,
        },
      ]
    }
    stack={false}
    style={Object {}}
  />
</ScrollablePlot>
`;

exports[`shows hint for mouse over: after mousing over first data point 1`] = `
<ScrollablePlot
  cnt={3}
>
  <XAxis
    attr="x"
    attrAxis="y"
    orientation="bottom"
  />
  <YAxis
    attr="y"
    attrAxis="x"
    orientation="left"
    tickFormat={[Function]}
  />
  <HorizontalGridLines
    attr="y"
    direction="horizontal"
  />
  <VerticalGridLines
    attr="x"
    direction="vertical"
  />
  <HorizontalBarSeries
    className=""
    color="#795548"
    data={
      Array [
        Object {
          "completed": 5,
          "completedCount": 1,
          "part": "abc",
          "totalCount": 11,
          "totalPlan": 10,
          "x": 10,
          "y": 1,
        },
        Object {
          "completed": 1,
          "completedCount": 6,
          "part": "def",
          "totalCount": 9,
          "totalPlan": 1,
          "x": 11,
          "y": 2,
        },
        Object {
          "completed": 0,
          "completedCount": 0,
          "part": "zzz",
          "totalCount": 6,
          "totalPlan": 15,
          "x": 16,
          "y": 3,
        },
      ]
    }
    onValueMouseOut={[Function]}
    onValueMouseOver={[Function]}
    stack={false}
    style={Object {}}
  />
  <AbstractSeries
    animation={false}
    className=""
    customComponent={[Function]}
    data={
      Array [
        Object {
          "x": 16,
          "y": 1,
        },
        Object {
          "x": 17,
          "y": 2,
        },
        Object {
          "x": 22,
          "y": 3,
        },
      ]
    }
    stack={false}
    style={Object {}}
  />
  <Hint
    align={
      Object {
        "horizontal": "auto",
        "vertical": "auto",
      }
    }
    format={[Function]}
    style={Object {}}
    value={
      Object {
        "completed": 5,
        "completedCount": 1,
        "part": "abc",
        "totalCount": 11,
        "totalPlan": 10,
        "x": 10,
        "y": 1,
      }
    }
  />
</ScrollablePlot>
`;
